### PR TITLE
Handle Keyboard Functionality

### PR DESCRIPTION
**Pull request by knittingdev**
Required for https://www.pivotaltracker.com/story/show/127652899

### CODE REVIEW

File: addon/mixins/cancelable.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    this.setProperties({
+      initialStartDate: startDate.clone(),
+      initialEndDate: endDate.clone(),
+      initialStartMonth: startMonth.clone(),
+      initialEndMonth: endMonth.clone()
+    });
+  },
+
+  didInsertElement() {
+    this.resetInitialValues();
```
Comment (iterion) [inline addon/mixins/cancelable.js]: Setting values in didInsertElement is deprecated: http://emberjs.com/deprecations/v1.x/#toc_modifying-a-property-within-code-didinsertelement-code
didReceiveAttrs is probably the more appropriate hook.

File: addon/mixins/picker.js [hunk 2]
```diff
@@ -8,17 +9,26 @@ const {
...
   endDate: moment().startOf('date'),
+  startMonth: moment().startOf('month'),
+  endMonth: moment().startOf('month'),
+
+  dropdownController: Ember.Object.create({
+    isOpen: false,
+  }),
+  initiallyOpened: false,
 
   _startDateToMoment: on('init', observer('startDate', function() {
```
Comment (iterion) [inline addon/mixins/picker.js]: Kill all observers. I know you're not responsible for this, so no need to fix now. But, I think we could probably just move what we're doing here to `didReceiveAttrs` as well.

File: addon/pods/date-range-picker/component.js [hunk 3]
```diff
@@ -65,10 +35,103 @@ export default Component.extend(Picker, Clearable, PickerActions, EKMixin, Click
...
+    let selectedIndex = this.selectedPresetIndex();
+    let currentPreset = this.get('presets')[selectedIndex];
+    if (currentPreset) {
+      Ember.set(currentPreset, "isSelected", false);
+    }
+    return [selectedIndex, currentPreset];
+  },
+
+  incrementSelectionIndex(currentPreset, currentIndex) {
+    if (currentIndex < 0 && !this.get('cancelSelected') && !this.get('applySelected')) { // just opened window
```
Comment (kluzny) [inline addon/pods/date-range-picker/component.js]: This top conditional may be met by the else conditional, if there is no further collision in the elseif's

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ember-date-range-picker",
-  "version": "0.2.0",
+  "version": "0.2.1",
```
Comment (kluzny) [inline package.json]: Does this upgrade into existing install with no changes? Might be a 0.3.0

File: tests/integration/pods/components/year-picker/component-test.js [hunk 5]
```diff
@@ -13,10 +12,12 @@ test('it renders', function(assert) {
...
 
-  this.render(hbs`{{year-picker start=start
-                                end=end}}`);
+  this.render(hbs`{{year-picker startDate=start
+                                endDate=end
+                                initiallyOpened=true}}`);
 
-  assert.equal(this.$().text().trim(), '');
+  let text = this.$().text().trim();
+  assert.equal(text.match(new RegExp('[0-9]{4}', 'g')).length, 21, 'Has 72 1-4 digit numbers...');
```
Comment (kluzny) [inline tests/integration/pods/components/year-picker/component-test.js]: "Has 72 1-4 digit numbers"!?

File: addon/pods/date-range-picker/template.hbs [hunk 6]
```diff
@@ -1,46 +1,51 @@
 <div class="dp-date-range-picker">
```
Comment (JeffPhan) [inline addon/pods/date-range-picker/template.hbs]: Can we start using `classNames` instead of adding another div?

Comment (kluzny) [inline addon/pods/date-range-picker/template.hbs]: I thought this too, but i didn't want to muddy the waters of this pr. In all fairness this line of code was never touched.

File: addon/pods/month-picker/template.hbs [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
+                        onKeydown=(action "handleKeydown")}}
+      {{range-input-mask value=rangeFormatted
+                         mask="9[9]/99[99]—9[9]/99[99]"
+                         class="dp-date-input"
+                         key-press="parseInput"}}
+    {{/dropdown.trigger}}
+  
+    {{#dropdown.content class="dp-panel"
+                        to="dp-month-picker"}}
       <div class="dp-display-month-year">
```
Comment (JeffPhan) [inline addon/pods/month-picker/template.hbs]: `classNames` could probably also be applied here as well as the other types of pickers

**Pull request merged**